File: swift/exercises.swift
import Foundation

struct NegativeAmountError: Error {}
struct NoSuchFileError: Error {}

func change(_ amount: Int) -> Result<[Int:Int], NegativeAmountError> {
    if amount < 0 {
        return .failure(NegativeAmountError())
    }
    var (counts, remaining) = ([Int:Int](), amount)
    for denomination in [25, 10, 5, 1] {
        (counts[denomination], remaining) = 
            remaining.quotientAndRemainder(dividingBy: denomination)
    }
    return .success(counts)
}

/**
 * Finds the first string in the array that matches the given predicate and converts it to lowercase.
 *
 * - Parameters:
 *   - of: The array of strings to search.
 *   - satisfying: The condition to match.
 * - Returns: The first matching string in lowercase, or nil if no match is found.
 */
func firstThenLowerCase(of strings: [String], satisfying predicate: (String) -> Bool) -> String? {
    return strings.first(where: predicate)?.lowercased()
}

/**
 * A class that represents a phrase and allows chaining additional phrases.
 *
 * - Property phrase: The initial phrase.
 */
class Say {
    private var _phrase: String

    var phrase: String {
        return _phrase
    }

    init(phrase: String) {
        self._phrase = phrase
    }

    /**
     * Chains the next phrase to the current phrase.
     *
     * - Parameter nextPhrase: The phrase to be added.
     * - Returns: A new Say instance with the combined phrase.
     */
    func and(_ nextPhrase: String) -> Say {
        return Say(phrase: "\(_phrase) \(nextPhrase)")
    }
}

/**
 * Creates a Say instance with the given phrase.
 *
 * - Parameter phrase: The initial phrase (default is an empty string).
 * - Returns: A Say instance.
 */
func say(_ phrase: String = "") -> Say {
    return Say(phrase: phrase)
}

/**
 * Counts the number of meaningful lines in a file.
 * A meaningful line is one that is not empty, not made up entirely of whitespace, and does not start with '#'.
 *
 * - Parameter fileName: The name of the file to read.
 * - Returns: A Result containing the number of meaningful lines or an error.
 */
func meaningfulLineCount(_ fileName: String) async -> Result<Int, NoSuchFileError> {
    guard let contents = try? String(contentsOfFile: fileName) else {
        return .failure(NoSuchFileError())
    }
    return .success(contents.split(separator: "\n").filter { line in
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        return !trimmed.isEmpty && !trimmed.hasPrefix("#")
    }.count)
}

/**
 * A struct representing a quaternion, which is a mathematical object used in three-dimensional
 * rotations and complex number extensions. This struct ensures immutability and provides
 * essential operations for quaternion arithmetic.
 *
 * Each quaternion is represented by four coefficients: a, b, c, and d, corresponding to
 * the components 1, i, j, and k respectively. The struct includes methods for addition,
 * multiplication, conjugation, and obtaining the coefficients as an array.
 *
 * Static constants ZERO, I, J, and K are provided for convenience, representing the
 * zero quaternion and the unit quaternions along the i, j, and k axes.
 *
 * - Properties:
 *   - a: The scalar part of the quaternion.
 *   - b: The coefficient of the i component.
 *   - c: The coefficient of the j component.
 *   - d: The coefficient of the k component.
 *
 * - Methods:
 *   - init(a:b:c:d:): Initializes a new quaternion with the given coefficients.
 *   - conjugate: Returns the conjugate of the quaternion.
 *   - coefficients: Returns the coefficients of the quaternion as an array.
 *   - description: Returns a string representation of the quaternion.
 *
 * - Operators:
 *   - static func +(_:_:): Adds two quaternions.
 *   - static func *(_:_:): Multiplies two quaternions.
 *
 * - Static Constants:
 *   - ZERO: The zero quaternion.
 *   - I: The unit quaternion along the i axis.
 *   - J: The unit quaternion along the j axis.
 *   - K: The unit quaternion along the k axis.
 */
struct Quaternion: CustomStringConvertible, Equatable {
    let a, b, c, d: Double

    init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.a = a
        self.b = b
        self.c = c
        self.d = d
    }

    static let ZERO = Quaternion(a: 0.0, b: 0.0, c: 0.0, d: 0.0)
    static let I = Quaternion(a: 0.0, b: 1.0, c: 0.0, d: 0.0)
    static let J = Quaternion(a: 0.0, b: 0.0, c: 1.0, d: 0.0)
    static let K = Quaternion(a: 0.0, b: 0.0, c: 0.0, d: 1.0)

    static func +(lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        return Quaternion(a: lhs.a + rhs.a, b: lhs.b + rhs.b, c: lhs.c + rhs.c, d: lhs.d + rhs.d)
    }

    static func *(lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        return Quaternion(
            a: lhs.a * rhs.a - lhs.b * rhs.b - lhs.c * rhs.c - lhs.d * rhs.d,
            b: lhs.a * rhs.b + lhs.b * rhs.a + lhs.c * rhs.d - lhs.d * rhs.c,
            c: lhs.a * rhs.c - lhs.b * rhs.d + lhs.c * rhs.a + lhs.d * rhs.b,
            d: lhs.a * rhs.d + lhs.b * rhs.c - lhs.c * rhs.b + lhs.d * rhs.a
        )
    }

    var conjugate: Quaternion {
        return Quaternion(a: a, b: -b, c: -c, d: -d)
    }

    var coefficients: [Double] {
        return [a, b, c, d]
    }

    var description: String {
        let coefficients = [a, b, c, d]
        let suffixes = ["", "i", "j", "k"]
        var s = ""

        for (i, coefficient) in coefficients.enumerated() {
            if coefficient == 0.0 { continue }
            if coefficient < 0 {
                s.append("-")
            } else if !s.isEmpty {
                s.append("+")
            }
            if abs(coefficient) != 1.0 || i == 0 {
                s.append("\(abs(coefficient))")
            }
            s.append(suffixes[i])
        }
        return s.isEmpty ? "0" : s
    }
}

/**
 * An enum representing a binary search tree (BST).
 * It permits two cases: Empty and Node.
 */
enum BinarySearchTree: CustomStringConvertible {
    case empty
    indirect case node(BinarySearchTree, String, BinarySearchTree)

    /**
     * The size of the BST.
     * - Returns: The number of nodes in the BST.
     */
    var size: Int {
        switch self {
        case .empty:
            return 0
        case let .node(left, _, right):
            return 1 + left.size + right.size
        }
    }

    /**
     * Checks if the BST contains a given value.
     * - Parameter value: The value to search for.
     * - Returns: True if the value is found, false otherwise.
     */
    func contains(_ value: String) -> Bool {
        switch self {
        case .empty:
            return false
        case let .node(left, v, right):
            if value < v {
                return left.contains(value)
            } else if value > v {
                return right.contains(value)
            } else {
                return true
            }
        }
    }

    /**
     * Inserts a new value into the BST.
     * - Parameter value: The value to insert.
     * - Returns: A new BST with the value inserted.
     */
    func insert(_ value: String) -> BinarySearchTree {
        switch self {
        case .empty:
            return .node(.empty, value, .empty)
        case let .node(left, v, right):
            if value < v {
                return .node(left.insert(value), v, right)
            } else if value > v {
                return .node(left, v, right.insert(value))
            } else {
                return self
            }
        }
    }

    var description: String {
        switch self {
        case .empty:
            return "()"
        case let .node(left, value, right):
            return "(\(left)\(value)\(right))"
        }
    }
}

File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

/**
 * Applies a consumer function to the first item in an array that matches a given predicate.
 *
 * @template T - The type of items in the array.
 * @template U - The return type of the consumer function.
 * @param {T[]} items - The array of items to search through.
 * @param {(item: T) => boolean} predicate - A function that returns true for the desired item.
 * @param {(item: T) => U} consumer - A function to apply to the found item.
 * @returns {U | undefined} - The result of the consumer function if an item is found, otherwise undefined.
 */
export function firstThenApply<T, U>(
  items: T[], 
  predicate: (item: T) => boolean, 
  consumer: (item: T) => U
): U | undefined {
  const foundItem = items.find(item => predicate(item))
  return foundItem ? consumer(foundItem) : undefined;
}

/**
 * Generates an infinite sequence of powers of a given base.
 *
 * @param {bigint} base - The base number for generating powers.
 * @yields {bigint} - The next power of the base in the sequence.
 */
export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power;
  }
}

/**
 * Counts the number of meaningful lines in a file.
 * A meaningful line is defined as a non-empty line that does not start with a '#' character.
 *
 * @param {string} fileName - The name of the file to be read.
 * @returns {Promise<number>} - A promise that resolves to the number of meaningful lines in the file.
 */
export async function meaningfulLineCount(fileName: string): Promise<number> {
  const fileHandle = await open(fileName, 'r');
  let count = 0;

  try {
    for await (const line of fileHandle.readLines()) {
      if (line.trim() !== '' && !line.trim().startsWith('#')) {
        count++;
      }
    }
  } finally {
      await fileHandle.close();
  }
  return count;
}

/**
 * Represents a sphere with a specific radius.
 */
interface Sphere {
  readonly kind: "Sphere";
  readonly radius: number;
}

/**
 * Represents a box with specific dimensions.
 */
interface Box {
  readonly kind: "Box";
  readonly width: number;
  readonly length: number;
  readonly depth: number;
}

/**
 * A union type representing either a Sphere or a Box.
 */
export type Shape = Sphere | Box;

/**
 * Calculates the surface area of a given shape.
 *
 * @param {Shape} shape - The shape for which to calculate the surface area.
 * @returns {number} - The surface area of the shape.
 */
export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2);
    case "Box":
      return 2 * (shape.width * shape.length + shape.length * shape.depth + shape.depth * shape.width);
  }
}

/**
 * Calculates the volume of a given shape.
 *
 * @param {Shape} shape - The shape for which to calculate the volume.
 * @returns {number} - The volume of the shape.
 */
export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
    case "Box":
      return shape.width * shape.length * shape.depth;
  }
}

/**
 * Converts a shape to a string representation.
 *
 * @param {Shape} shape - The shape to convert to a string.
 * @returns {string} - A string representation of the shape.
 */
export function shapeToString(shape: Shape): string {
  switch (shape.kind) {
    case "Sphere":
      return `Sphere with radius ${shape.radius}`;
    case "Box":
      return `Box with width ${shape.width}, height ${shape.length}, and depth ${shape.depth}`;
  }
}

/**
 * Compares two shapes for equality.
 *
 * @param {Shape} shape1 - The first shape to compare.
 * @param {Shape} shape2 - The second shape to compare.
 * @returns {boolean} - True if the shapes are equal, false otherwise.
 */
export function equals(shape1: Shape, shape2: Shape): boolean {
  if (shape1.kind !== shape2.kind) {
    return false;
  }
  if (shape1.kind === "Sphere" && shape2.kind === "Sphere") {
    return shape1.radius === shape2.radius;
  }
  if (shape1.kind === "Box" && shape2.kind === "Box") {
    return shape1.width === shape2.width &&
           shape1.length === shape2.length &&
           shape1.depth === shape2.depth;
  }
  return false;
}

/**
 * Interface representing a generic Binary Search Tree (BST).
 */
export interface BinarySearchTree<T> {
  /**
   * Returns the number of elements in the BST.
   * @returns {number} The size of the BST.
   */
  size(): number;

  /**
   * Inserts a value into the BST and returns the updated BST.
   * @param {T} value - The value to insert.
   * @returns {BinarySearchTree<T>} The updated BST.
   */
  insert(value: T): BinarySearchTree<T>;

  /**
   * Checks if a value is present in the BST.
   * @param {T} value - The value to check.
   * @returns {boolean} True if the value is present, false otherwise.
   */
  contains(value: T): boolean;

  /**
   * Returns an iterable for in-order traversal of the BST.
   * @returns {Iterable<T>} An iterable for in-order traversal.
   */
  inorder(): Iterable<T>;

  /**
   * Returns a string representation of the BST.
   * @returns {string} The string representation of the BST.
   */
  toString(): string;
}

/**
 * Class representing an empty node in a BST.
 */
export class Empty<T> implements BinarySearchTree<T> {
  /**
   * Returns the number of elements in the BST.
   * @returns {number} The size of the BST, which is 0 for an empty node.
   */
  size(): number {
    return 0;
  }

  /**
   * Inserts a value into the BST and returns a new node containing the value.
   * @param {T} value - The value to insert.
   * @returns {BinarySearchTree<T>} A new node containing the value.
   */
  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty());
  }

  /**
   * Checks if a value is present in the BST.
   * @param {T} value - The value to check.
   * @returns {boolean} False, as an empty node cannot contain any value.
   */
  contains(value: T): boolean {
    return false;
  }

  /**
   * Returns an iterable for in-order traversal of the BST.
   * @returns {Iterable<T>} An empty iterable.
   */
  inorder(): Iterable<T> {
    return this._inorder();
  }

  /**
   * Private generator method for in-order traversal.
   * @returns {Iterable<T>} An empty iterable.
   */
  private *_inorder(): Iterable<T> {
    return;
  }

  /**
   * Returns a string representation of the empty node.
   * @returns {string} The string representation of the empty node, which is "()".
   */
  toString(): string {
    return "()";
  }
}

/**
 * Class representing a node in a BST.
 */
class Node<T> implements BinarySearchTree<T> {
  private value: T;
  private left: BinarySearchTree<T>;
  private right: BinarySearchTree<T>;

  /**
   * Creates a new node.
   * @param {T} value - The value of the node.
   * @param {BinarySearchTree<T>} left - The left subtree.
   * @param {BinarySearchTree<T>} right - The right subtree.
   */
  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    this.value = value;
    this.left = left;
    this.right = right;
  }

  /**
   * Returns the number of elements in the BST.
   * @returns {number} The size of the BST.
   */
  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  /**
   * Inserts a value into the BST and returns the updated BST.
   * @param {T} value - The value to insert.
   * @returns {BinarySearchTree<T>} The updated BST.
   */
  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right);
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value));
    }
    return this;
  }

  /**
   * Checks if a value is present in the BST.
   * @param {T} value - The value to check.
   * @returns {boolean} True if the value is present, false otherwise.
   */
  contains(value: T): boolean {
    if (value === this.value) {
      return true;
    } else if (value < this.value) {
      return this.left.contains(value);
    } else {
      return this.right.contains(value);
    }
  }

  /**
   * Returns an iterable for in-order traversal of the BST.
   * @returns {Iterable<T>} An iterable for in-order traversal.
   */
  inorder(): Iterable<T> {
    return this._inorder();
  }

  /**
   * Private generator method for in-order traversal.
   * @returns {Iterable<T>} An iterable for in-order traversal.
   */
  private *_inorder(): Iterable<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  /**
   * Returns a string representation of the node.
   * @returns {string} The string representation of the node.
   */
  toString(): string {
    const leftStr = this.left instanceof Empty ? "" : this.left.toString();
    const rightStr = this.right instanceof Empty ? "" : this.right.toString();
    return `(${leftStr}${this.value}${rightStr})`;
  }
}